### ISSUE TITLE
path_to_parts parsing bug

### ISSUE THREAD

**Issue by kserhii** [sanic-org/sanic-routing#39]
**Describe the bug**

In my project, I have an endpoint

```python
@api.route(
    r'/image/iiif/'
    r'<image_id>/'
    r'<region:full|square|\d+,\d+,\d+,\d+>/'
    r'<size:max|\d+,|,\d+|\d+,\d+>/'
    r'<rotation:int>/'
    r'default.jpg')
```
It works for Sanic==20.12.1, but after migration to Sanic==21.3.4 it's not a valid router anymore.

```
parts = ('api', 'image', 'iiif', '<image_id>/<region:full|square|\\d+,\\d+,\\d+,\\d+>', '<size:max|\\d+,|,\\d+|\\d+,\\d+>', '<rotation:int>', ...)
delimiter = '/'

    def parts_to_path(parts, delimiter="/"):
        path = []
        for part in parts:
            if part.startswith("<"):
                try:
                    match = REGEX_PARAM_NAME.match(part)
                    param_type = ""
>                   if match.group(2):
E                   AttributeError: 'NoneType' object has no attribute 'group'

../venv/lib/python3.8/site-packages/sanic_routing/utils.py:76: AttributeError
```

**How to reproduce**

```python
>>> from sanic_routing.utils import path_to_parts
>>> s = (
...     r'/image/iiif/'
...     r'<image_id>/'
...     r'<region:full|square|\d+,\d+,\d+,\d+>/'
...     r'<size:max|\d+,|,\d+|\d+,\d+>/'
...     r'<rotation:int>/'
...     r'default.jpg')
>> path_to_parts(s)
('image', 'iiif', '<image_id>/<region:full|square|\\d+,\\d+,\\d+,\\d+>', '<size:max|\\d+,|,\\d+|\\d+,\\d+>', '<rotation:int>', 'default.jpg')
```
**Expected behavior**
```
>> path_to_parts(s)
('image', 'iiif', '<image_id>', '<region:full|square|\\d+,\\d+,\\d+,\\d+>', '<size:max|\\d+,|,\\d+|\\d+,\\d+>', '<rotation:int>', 'default.jpg')
```
**Environment**

- OS: Linux (Ubuntu 20.04)
- Python: 3.8
- Sanic version: 21.3.4
- Sanic-routing version: 0.6.2

**Comment by ahopkins** [comment]
Hmm, I think it might be worth it to scrap the regex splitting in this function. I have tested it quite a bit, but it does seem like this is a use case that breaks it. Perhaps a solution will be to run a regular `.split("/")` operation, and then loop through the segments looking for instances where there is an open `<` but the close `>` is in the next segment, and merge.

**Comment by ahopkins** [comment]
```python
def _normalize(parts, delimiter):
    part_iter = iter(parts)
    normalized = []
    for part in part_iter:
        if part.startswith("<") and not part.endswith(">"):
            ending = ""
            try:
                while not ending.endswith(">"):
                    ending += delimiter + next(part_iter)
            except StopIteration:
                raise Exception("invalid")

            part += ending
        normalized.append(part)
    return normalized


def path_to_parts(path, delimiter="/"):
    parts = _normalize(path.split(delimiter), delimiter)
    return tuple([part if part.startswith("<") else quote(part) for part in parts])
```

I just threw together this little snippet to prove my point. I think this is probably the way to go and should be more accurate in catching corner cases like this one.

**Comment by kserhii** [comment]
Can you maybe provide a `path` for which normal `.split("/")` does not work? Just to have a clear understanding of the problem.

**Comment by ahopkins** [comment]
> Can you maybe provide a `path` for which normal `.split("/")` does not work? Just to have a clear understanding of the problem.

```python
@app.get(r"/path/to/<deeply_nested:[a-z/]+>/thing")
async def handler(request, deeply_nested):
    return text(deeply_nested)
```

```bash
$ curl localhost:9999/path/to/location/with/slashes/thing
location/with/slashes
```

**Comment by kserhii** [comment]
I've looked a bit and found the solution her:
https://stackoverflow.com/questions/732029/how-to-split-string-by-unless-is-within-brackets-using-regex

Here is the [pull request](https://github.com/sanic-org/sanic-routing/pull/40) with extra tests

**Comment by ahopkins** [comment]
Fixed by #40
